VersionControlUI/src/DiffManager.cpp

```diff
@@ -0,0 +1,300 @@
+			auto items = diffViewItem->findAllVisualizationsOf(oldNode);
+			items.append(diffViewItem->findAllVisualizationsOf(newNode));
+			for (auto item : items)
+			{
+				auto overlay = new Visualization::HighlightOverlay{item,
+						Visualization::HighlightOverlay::itemStyles().get("move_no_bg_solid")};
+				item->addOverlay(overlay, "move_highlights");
+			}});
+	}
+
+	// create highlights for deleted nodes
+	for (auto id : nodeLists[INDEX_OF_DELETE])
+	{
+		auto oldNode = const_cast<Model::Node*>(oldVersionManager->nodeIdMap().node(id));
+
+		if (!oldNode)
+			continue;
+
+		 Visualization::VisualizationManager::instance().mainScene()->addPostEventAction(
+								[oldNode, diffViewItem](){
+			 auto oldItems =  diffViewItem->findAllVisualizationsOf(oldNode);
+			 for (auto item : oldItems)
+			 {
+				 auto overlay = new Visualization::HighlightOverlay{item,
+						 Visualization::HighlightOverlay::itemStyles().get("delete_no_bg_solid")};
+				 item->addOverlay(overlay, "delete_highlights");
+			 }});
+
+	}
+
+	// create highlights for added nodes
+	for (auto id : nodeLists[INDEX_OF_ADD])
+	{
+		auto newNode = const_cast<Model::Node*>(newVersionManager->nodeIdMap().node(id));
+
+		Visualization::VisualizationManager::instance().mainScene()->addPostEventAction(
+							  [newNode, diffViewItem](){
+			auto newItems =  diffViewItem->findAllVisualizationsOf(newNode);
+			for (auto item : newItems)
+			{
+				auto overlay = new Visualization::HighlightOverlay{item,
+						Visualization::HighlightOverlay::itemStyles().get("insert_no_bg_solid")};
+				item->addOverlay(overlay, "insert_highlights");
+			}});
+	}
+
+	// create highlights for modified nodes
+	for (auto id : nodeLists[INDEX_OF_MODIFY])
+	{
+		auto oldNode = const_cast<Model::Node*>(oldVersionManager->nodeIdMap().node(id));
+		auto newNode = const_cast<Model::Node*>(newVersionManager->nodeIdMap().node(id));
+
+		Visualization::VisualizationManager::instance().mainScene()->addPostEventAction(
+							  [oldNode, newNode, diffViewItem](){
+			auto items = diffViewItem->findAllVisualizationsOf(oldNode);
+			items.append(diffViewItem->findAllVisualizationsOf(newNode));
+			for (auto item : items)
+			{
+				auto overlay = new Visualization::HighlightOverlay{item,
+						Visualization::HighlightOverlay::itemStyles().get("modify_no_bg_solid")};
+				item->addOverlay(overlay, "modify_highlights");
+			}});
+	}
+
+	// switch to the newly created view
+	Visualization::VisualizationManager::instance().mainScene()->viewItems()->switchToView(diffViewItem);
+}
+
+/**
+ * Deletes all nodes in \a container if their direct parent is also in \container.
+ */
+template<typename Container>
+void DiffManager::removeDirectChildrenOfNodesInContainer(Container* container)
+{
+	auto it = container->begin();
+	while (it != container->end())
+	{
+		auto nodeInContainer = Model::AllTreeManagers::instance().nodeForId(*it);
+		auto parentNode = nodeInContainer->parent();
```

Comment:
This algorithm is only correct, if you guarantee that you explore nodes from the leaves towards the root. Otherwise
if A, B, and C are in the container and A is a parent of B is a parent of C, you might first explore B and then remove it because A is in there, and then if you explore C you will not remove it.